### DISCUSSION TITLE
react hook form vs formik

### DISCUSSION THREAD

**Discussion by ahmedosama5200** [react-hook-form/react-hook-form#3004] [General] [answered]
Hi everyone
I have been using formik for a while now and I am thinking about migrating to react hook form. The thing is, React official documentation recommends using [controlled components](https://reactjs.org/docs/forms.html), even though they are less performant than using refs. Why is that ?
Formik does exactly what react recommends while react hook form goes against it.

**Comment by bluebill1049** [comment, marked as answer]
short answer: 
we support both uncontrolled/controlled. more to read here: 

https://react-hook-form.com/faqs#ReactHookFormFormikorReduxForm

<img width="961" alt="Screen Shot 2020-09-26 at 10 40 07 am" src="https://user-images.githubusercontent.com/10513364/94325992-a5f2ca80-ffe4-11ea-9238-44b0633dfe93.png">

We are not aginst react recommendation for building **component** controlled, however building forms over the years and I start to realize the native (uncontrolled) form is just fine or even better in the most use cases. especially simplicity, performance, and scalable for large and complex forms. now you still can build controlled inputs at React Hook Form by leveraging Controller: https://react-hook-form.com/api#Controller There are cases where controlled forms make sense, but I find less the case, more of a controlled component.

**Comment by NikolaGrujicic** [comment]
I  have used recently both libraries on a project, some of the important things to consider when choosing a form library are( personally I prefer to use react-hook-form ):
  - Number of stars on GitHub and weekly downloads on NPM
  - Performance
  - Development - Is it well maintained
  - Customisation and extendibility
  - Size of the library
  - Open issues
  - Updates - how often is it updated
If you would like to get more deep into this topic you can check out this blog post where the two libraries are compared in detail: https://www.framelessgrid.com/react-hook-form-vs-formik-react-form-library-comparison-in-2021/

**Reply by bluebill1049** [reply]
Thank you very much @NikolaGrujicic for the kind words and the blog post! ❤️

**Reply by NikolaGrujicic** [reply]
> Thank you very much @NikolaGrujicic for the kind words and the blog post! ❤️

Thank you for your nice reply and for making this awesome library, if you have time please share my post on any social media account you have, thank you again, and best regards ❤️

**Reply by bluebill1049** [reply]
Will do 👍

**Comment by behzadmehrabi** [comment]
one of my concerns for migrating fully to `react-hook-form` is that I don't know how it's gonna end up in the long term.
at the time being @bluebill1049 is putting so much time and effort into `react-hook-form` which is awesome, but is he gonna keep doing it like this for years? I think teams/companies have to become a sponsor of this project, It will really help out with the continued development of the `react-hook-form`.

**Reply by bluebill1049** [reply]
This is definitely a problem of open source in general. Right now I spend a lot of my free time working on this library, and core team members are slowly walking away because it is very little or 0 to compensate for their time investment. I hope once day more business will willing to contribute to open source.

**Reply by msageryd** [reply]
As a converter from Final-Form I'd like to chime in. A little late, I know, but the subject is important..

Final-Form has served me very well. It's a great library. But with only one maintainer it seems unevitable for a project to slow down at some point. Final-Form has definitely stagnated, it's now almost impossible to get answers from Erik. I don't blame him at all. He has done extraordinary work for us freeloaders. OSS is hard.

After realising that FF had stagnated to almost full stop, I started to investigate Formik. Jared is a very sharp mind and Formik is a great library. But he seems to have left the project and puts his time into Vercel. Unfortunately he seems to have taken the maintenance keys with him, and people are struggling to get stuff merged. Version 3 was anounced a year ago, and still hasn't been merged. It lives in a fork if I've understood correctly.

The same risk of stagnation applies to RHF as well. But so far it seems very lively. I have never seen an issue-list clean as that of RHF. Thousands of closed issues and zero or only a couple of open issues.

These three libraries are the top ones in my book. The stagnation risk is about the same for all of them, so I'm calling this a tie in my evaluation.

This made me take a decision based on the library itself, the DX and UX and the overall structure. I must say that I'm impressed with RHS in this regard. I'm still very new to the library, but with very little effort I have now converted most of my FF stuff to RHF. Some quirks remains (on my side), but they will soon be solved.

To mitigate the stagnation risk somewhat I have become an RHF sponsor via OpenCollective.

Hopefully RHS will be reasonably maintained going forward. And if Bill should loose his spark or get too much to do on other projects, I can only hope that he will welcome other maintainers and also give them the keys.

edit: this was ment as a reply to @Bezmehrabi , but I managed to reply to @bluebill1049 instead..
